hazelcast/src/main/java/com/hazelcast/internal/server/tcp/CompatibilityTcpServerControl.java

```diff
@@ -0,0 +1,209 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.server.tcp;
+
+import com.hazelcast.cluster.Address;
+import com.hazelcast.instance.EndpointQualifier;
+import com.hazelcast.instance.ProtocolType;
+import com.hazelcast.internal.compatibility.cluster.impl.CompatibilityExtendedBindMessage;
+import com.hazelcast.internal.compatibility.nio.tcp.CompatibilitySendMemberHandshakeTask;
+import com.hazelcast.internal.nio.Connection;
+import com.hazelcast.internal.nio.ConnectionType;
+import com.hazelcast.internal.nio.Packet;
+import com.hazelcast.internal.nio.Packet.Type;
+import com.hazelcast.internal.server.ServerContext;
+import com.hazelcast.logging.ILogger;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+
+import static com.hazelcast.spi.properties.ClusterProperty.CHANNEL_COUNT;
+
+/**
+ * Handler responsible for processing bind messages from 3.x members.
+ */
+public final class CompatibilityTcpServerControl {
+
+    private final TcpServerConnectionManager connectionManager;
+    private final ServerContext serverContext;
+    private final ILogger logger;
+    private final boolean unifiedEndpointManager;
+    private final Set<ProtocolType> supportedProtocolTypes;
+    private final int expectedPlaneCount;
+
+    public CompatibilityTcpServerControl(TcpServerConnectionManager connectionManager,
+                                         ServerContext serverContext, ILogger logger,
+                                         Set<ProtocolType> supportedProtocolTypes) {
+        this.connectionManager = connectionManager;
+        this.serverContext = serverContext;
+        this.logger = logger;
+        this.supportedProtocolTypes = supportedProtocolTypes;
+        this.unifiedEndpointManager = connectionManager.getEndpointQualifier() == null;
+        this.expectedPlaneCount = serverContext.properties().getInteger(CHANNEL_COUNT);
+    }
+
+    public void process(Packet packet) {
+        if (packet.getPacketType() == Type.COMPATIBILITY_BIND_MESSAGE) {
+            logger.info("Ignoring compatibility bind message " + packet);
+            return;
+        }
+        CompatibilityExtendedBindMessage handshake = serverContext.getCompatibilitySerializationService().toObject(packet);
+        TcpServerConnection connection = (TcpServerConnection) packet.getConn();
+
+        if (!connection.setHandshake()) {
+            if (logger.isFinestEnabled()) {
+                logger.finest("Connection " + connection + " handshake is already completed, ignoring incoming " + handshake);
+            }
+            return;
+        }
+
+        if (handshake.getPlaneCount() != expectedPlaneCount) {
```

Comment:
I think we should check against a constant 1, since for the compatibility connections we don't support multiple connections between the members. Here we check against the `CHANNEL_COUNT` property, which is then shared with the regular member-to-member connections, hence, setting it to non-default will break the migration connection. Maybe this is something we want to explicitly test.

Reply:
Looks like there's an issue where the 4.x cluster sets a plane count > 1. Here's an example if we're replicating 3.x -> 4.x:

1. 3.x member establishes a connection with 4.x, does the handshake, connection ends up in plane 0. 3.x member sends a WAN batch over that connection.
2. 4.x member tries to send the response, it [chooses a random plane](https://github.com/hazelcast/hazelcast/blob/ab91141e18b80a1724e64e141432fd8a0a6ad1e0/hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OutboundResponseHandler.java#L214-L219), which does not need to be zero. If then tries to initiate connection by sending a `MemberHandshake`, since it does not know if it’s talking to 3.x or 4.x
3. a) If the other side is a plain 3.x member, you get a deserialisation exception, response fails to be sent, errors, whatnot
3. b) if the other side is a migration 3.x member, it deserialises it to a CompatibilityBindMessage. It will then reply with a CompatibilityBindMessage, but it does not know anything about plane counts so it doesn’t serialize that options map
4. the 4.x member deserializes the response as MemberHandshake, but it sees the sender didn’t send the options map, so it uses the RU code you’ve sent and sets the defaults
5. the 4.x member compares the plane count of the received handshake and the local count, and sees they are different. Again, exception.

@blazember and I detected a similar issue can happen when there's a RU from a 4.0 member to a 4.1 member. In this scenario, the 4.0 member would fail in the same way as a 3.12 member.

We have two options I'm looking into:
• when we do a compatibility handshake on a 4.x member, just put the same connection in all planes so any response will find the same connection
• say migration tool does not work with planeCount>1

Reply:
Set the expected count to 1. Although, as discussed and mentioned above, there's an issue why migration tools don't support multi-plane connections. If we decide to fix it, I can make a test which will then pass. I used to have a test, but it randomly fails because of the issue above.